src/core/pointcloud/qgspointcloudstatscalculator.cpp

```diff
@@ -0,0 +1,268 @@
+ ***************************************************************************/
+
+#include "qgspointcloudstatscalculator.h"
+
+#include "qgspointcloudindex.h"
+#include "qgsmessagelog.h"
+#include "qgspointcloudattribute.h"
+#include "qgspointcloudrequest.h"
+
+#include "qgspointcloudrenderer.h"
+
+#include "qgsapplication.h"
+#include "qgspointcloudstatscalculationtask.h"
+#include "qgsfeedback.h"
+#include "qgspointcloudblockrequest.h"
+
+#include <QQueue>
+#include <QtConcurrent/QtConcurrentMap>
+
+struct StatsProcessor
+{
+    typedef QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> result_type;
+
+    StatsProcessor( QgsPointCloudIndex *index, QgsPointCloudRequest request, QgsFeedback *feedback )
+      : mIndex( index ), mRequest( request ), mFeedback( feedback )
+    {
+
+    }
+
+    QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> operator()( IndexedPointCloudNode node )
+    {
+      QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> statsMap;
+      for ( QgsPointCloudAttribute attribute : mRequest.attributes().attributes() )
+      {
+        QgsPointCloudStatsCalculator::AttributeStatistics summary;
+        summary.minimum = std::numeric_limits<double>::max();
+        summary.maximum = std::numeric_limits<double>::lowest();
+        summary.count = 0;
+        summary.classCount.clear();
+        statsMap[ attribute.name() ] = summary;
+      }
+
+      std::unique_ptr<QgsPointCloudBlock> block;
+      if ( mIndex->accessType() == QgsPointCloudIndex::Local )
+      {
+        block.reset( mIndex->nodeData( node, mRequest ) );
+      }
+      else
+      {
+        QgsPointCloudBlockRequest *request = mIndex->asyncNodeData( node, mRequest );
+        QEventLoop loop;
+        QObject::connect( request, &QgsPointCloudBlockRequest::finished, &loop, &QEventLoop::quit );
+        QObject::connect( mFeedback, &QgsFeedback::canceled, &loop, &QEventLoop::quit );
+        loop.exec();
+        if ( !mFeedback->isCanceled() )
+          block.reset( request->block() );
+        if ( !request->block() )
+        {
+          QgsMessageLog::logMessage( QObject::tr( "Unable to calculate statistics for node %1, error: \"%2\"" ).arg( node.toString() ).arg( request->errorStr() ) );
+          return result_type();
+        }
+      }
+
+      if ( !block.get() )
+      {
+        return result_type();
+      }
+
+      const char *ptr = block->data();
+      int count = block->pointCount();
+      const QgsPointCloudAttributeCollection attributes = block->attributes();
+      int recordSize = attributes.pointRecordSize();
+
+      for ( int i = 0; i < count; ++i )
+      {
+        for ( QgsPointCloudAttribute attribute : attributes.attributes() )
+        {
+          if ( mFeedback->isCanceled() )
+          {
+            return result_type();
+          }
+          double attributeValue = 0;
+          int attributeOffset = 0;
+          attributes.find( attribute.name(), attributeOffset );
```

Comment:
It may be more efficient to create a vector of attribute offsets before going through the points, so that we don't need to look up attribute offsets again and again many times.